File: src/web-app/src/api/backend.ts
import type { Connection, Drive, FileContent, Operation, PanelState, Side } from './types'

/**
 * The single seam between the web UI and whatever fulfills it.
 *
 * Components never talk to a transport: no fetch, no WebSocket, no URLs. They
 * call command methods (promises) and subscribe to pushed events (callbacks).
 * How commands travel and where events come from is an implementation detail:
 *
 * `HttpBackend` (api/http.ts) is the implementation: REST requests and WebSocket
 * streams to an ice-commander instance running with `--webui`.
 *
 * Keep this interface in lockstep with the app's REST/WS contract (see
 * src/gtk-app/src/api.rs); it is the frontend mirror of that contract.
 */

/** Server-pushed events. Register via [`Backend.subscribe`]; unused ones may be omitted. */
export interface BackendEvents {
  /** A panel changed (navigation, mutation, or a mirrored GTK action) — apply the new state. */
  onPanelUpdated?(side: Side, state: PanelState): void
  /** The GTK side opened/closed a panel terminal — mirror it. */
  onTerminalState?(side: Side, open: boolean): void
  /** A panel terminal was expanded across both panels (or collapsed back). */
  onTerminalExpanded?(side: Side, expanded: boolean): void
  /** The app asks the web UI to open its file viewer (remote F3/F4). */
  onOpenViewer?(side: Side, path: string, mode: 'view' | 'edit'): void
  /** A native viewer window closed in the GTK app — close the web overlay to match. */
  onCloseViewer?(side: Side): void
  /** The app opened or closed its own connections dialog — mirror the web overlay. */
  onConnectionsDialog?(open: boolean): void
  /** The GTK side switched a panel's file view ("list" | "small" | "large"). */
  onViewMode?(side: Side, mode: string): void
}

/** Callbacks for one live terminal. `onData` receives raw PTY output (VT100 bytes for xterm). */
export interface TerminalCallbacks {
  onData(data: Uint8Array): void
  /** The session ended on the far side (socket closed, shell exited). */
  onClose?(): void
}

/** A live terminal session. Keystrokes go in via `send`; output arrives on the callbacks. */
export interface TerminalSession {
  send(data: string): void
  /** Tell the far-side PTY the xterm dimensions, so full-screen TUIs (mc/vim) fill the view. */
  resize(cols: number, rows: number): void
  close(): void
}

export interface Backend {
  // ── panel state & navigation ────────────────────────────────────────────────
  fetchState(side: Side): Promise<PanelState>
  enter(side: Side, name: string): Promise<void>
  goUp(side: Side): Promise<void>
  goBack(side: Side): Promise<void>
  goForward(side: Side): Promise<void>
  /** Jump to a breadcrumb by its flat level index (0 = provider root). */
  breadcrumb(side: Side, level: number): Promise<void>
  /** Show the Select Source page in this panel. */
  goHome(side: Side): Promise<void>

  // ── tabs ──────────────────────────────────────────────────────────────────────
  /** Open a new tab (focused) on this side. */
  addTab(side: Side): Promise<void>
  /** Close the tab with `id` (no-op on the last tab). */
  closeTab(side: Side, id: number): Promise<void>
  /** Make the tab with `id` the active one. */
  switchTab(side: Side, id: number): Promise<void>

  // ── sources ─────────────────────────────────────────────────────────────────
  getDrives(): Promise<Drive[]>
  /** Open a source (drive / connection) by its unified key in the given panel. */
  activateSource(side: Side, key: string): Promise<void>
  getConnections(): Promise<Connection[]>
  /** Persist a connection (upsert by name) — it appears in getDrives() as kind 'net'
   *  and mounts via activateSource. Does not mount by itself. */
  saveConnection(connection: Connection): Promise<void>
  /** Remove a saved connection by name. */
  deleteConnection(name: string): Promise<void>
  setConnectionsDialog(open: boolean): Promise<void>
  exportConnections(password?: string): Promise<string>
  importConnections(data: string, password?: string): Promise<number>
  connectTo(side: Side, connection: Connection): Promise<void>

  // ── favorites ───────────────────────────────────────────────────────────────
  /** `key` is the unified drive key (e.g. "local_fs:/"). */
  toggleFavorite(key: string): Promise<void>
  getFavoritesOnly(): Promise<boolean>
  setFavoritesOnly(value: boolean): Promise<void>

  // ── files ───────────────────────────────────────────────────────────────────
  readFile(side: Side, path: string): Promise<FileContent>
  writeFile(side: Side, path: string, content: string): Promise<void>
  /** URL suitable for <img src>/<video src> — how the bytes arrive is the
   *  backend's business (REST stream endpoint, data URI, blob…). */
  getFileUrl(side: Side, path: string): Promise<string>
  /** Also open the file in the app's native GTK viewer window (mirror a web F3/F4). */
  openNative(side: Side, path: string): Promise<void>
  /** Close the app's native viewer windows (mirror the web overlay closing). */
  closeViewerWindows(): Promise<void>
  /** Raw-bytes upload (drag-and-drop from the OS). `path` is the full target
   *  file path in the panel's provider — works on remote mounts too. */
  uploadFile(side: Side, path: string, data: Blob): Promise<void>
  deleteEntries(side: Side, paths: string[]): Promise<void>
  mkdir(side: Side, name: string): Promise<void>
  renameEntry(side: Side, oldPath: string, newName: string): Promise<void>
  copyEntries(srcSide: Side, dstSide: Side, paths: string[]): Promise<void>
  moveEntries(srcSide: Side, dstSide: Side, paths: string[]): Promise<void>

  // ── operations (Processes popup) ────────────────────────────────────────────
  /** Snapshot of running transfers; the popup polls while open. */
  getOperations(): Promise<Operation[]>
  /** Abort a running transfer — same effect as its Cancel button. */
  cancelOperation(id: number): Promise<void>

  // ── terminal ────────────────────────────────────────────────────────────────
  /** Open a terminal for a panel. The transport (WS, fake shell, …) is hidden here. */
  openTerminal(side: Side, callbacks: TerminalCallbacks): TerminalSession
  /** Expand the panel's terminal across both panels (or collapse). Echoes back as an event. */
  setTerminalExpanded(side: Side, expanded: boolean): Promise<void>
  /** Switch the panel's file view ("list" | "small" | "large"). Echoes back as an event. */
  setViewMode(side: Side, mode: string): Promise<void>

  // ── events ──────────────────────────────────────────────────────────────────
  /** Start receiving pushed events. Returns an unsubscribe function. */
  subscribe(events: BackendEvents): () => void
}

File: src/panel-server/src/backend.rs
use crate::{
    ApiCmd, ApiConnection, ApiDrive, ApiFileContent, ApiPanelState, ApiResult, PanelSide,
};

#[async_trait::async_trait(?Send)]
pub trait PanelBackend {
    async fn enter(&self, side: PanelSide, name: String) -> ApiResult<()>;
    async fn go_up(&self, side: PanelSide) -> ApiResult<()>;
    async fn go_back(&self, side: PanelSide) -> ApiResult<()>;
    async fn go_forward(&self, side: PanelSide) -> ApiResult<()>;
    async fn go_to_level(&self, side: PanelSide, level: usize) -> ApiResult<()>;
    async fn go_home(&self, side: PanelSide) -> ApiResult<()>;

    async fn add_tab(&self, side: PanelSide) -> ApiResult<()>;
    async fn close_tab(&self, side: PanelSide, id: u32) -> ApiResult<()>;
    async fn switch_tab(&self, side: PanelSide, id: u32) -> ApiResult<()>;

    fn read_state(&self, side: PanelSide) -> ApiPanelState;

    async fn delete(&self, side: PanelSide, paths: Vec<String>) -> ApiResult<()>;
    async fn mkdir(&self, side: PanelSide, name: String) -> ApiResult<()>;
    async fn rename(&self, side: PanelSide, old_path: String, new_name: String) -> ApiResult<()>;
    async fn copy(&self, src: PanelSide, dst: PanelSide, paths: Vec<String>) -> ApiResult<()>;
    async fn move_entries(&self, src: PanelSide, dst: PanelSide, paths: Vec<String>) -> ApiResult<()>;
    async fn read_file(&self, side: PanelSide, path: String) -> ApiResult<ApiFileContent>;
    async fn stream_file(&self, side: PanelSide, path: String) -> ApiResult<Vec<u8>>;
    async fn write_file(&self, side: PanelSide, path: String, content: String) -> ApiResult<()>;
    async fn upload_file(&self, side: PanelSide, path: String, data: Vec<u8>) -> ApiResult<()>;

    fn get_drives(&self) -> Vec<ApiDrive>;
    async fn activate_source(&self, side: PanelSide, key: String) -> ApiResult<()>;
    fn get_connections(&self) -> Vec<ApiConnection>;
    fn save_connection(&self, connection: ApiConnection) -> ApiResult<()>;
    fn delete_connection(&self, name: String) -> ApiResult<()>;
    async fn connect_to(&self, side: PanelSide, connection: ApiConnection) -> ApiResult<()>;
    async fn refresh_panel(&self, side: PanelSide) -> ApiResult<()>;
    fn get_settings(&self) -> ApiResult<serde_json::Value>;
    fn set_settings(&self, values: serde_json::Value) -> ApiResult<()>;
    fn viewer_content(&self) -> ApiResult<Option<(String, String, String)>>;
    fn set_connections_dialog(&self, open: bool) -> ApiResult<()>;
    fn export_connections(&self, password: Option<String>) -> ApiResult<String>;
    fn import_connections(&self, data: String, password: Option<String>) -> ApiResult<usize>;

    fn toggle_favorite(&self, path: String);
    fn get_favorites_only(&self) -> bool;
    fn set_favorites_only(&self, value: bool);
}

pub async fn dispatch_core<B: PanelBackend + ?Sized>(backend: &B, cmd: ApiCmd) -> Option<ApiCmd> {
    match cmd {
        ApiCmd::Enter { side, name, reply } => {
            let _ = backend.enter(side, name).await;
            let _ = reply.send(Ok(()));
        }
        ApiCmd::GoUp { side, reply } => {
            let _ = backend.go_up(side).await;
            let _ = reply.send(Ok(()));
        }
        ApiCmd::GoBack { side, reply } => {
            let _ = backend.go_back(side).await;
            let _ = reply.send(Ok(()));
        }
        ApiCmd::GoForward { side, reply } => {
            let _ = backend.go_forward(side).await;
            let _ = reply.send(Ok(()));
        }
        ApiCmd::Breadcrumb { side, level, reply } => {
            let _ = backend.go_to_level(side, level).await;
            let _ = reply.send(Ok(()));
        }
        ApiCmd::GoHome { side, reply } => {
            let _ = reply.send(backend.go_home(side).await);
        }

        ApiCmd::AddTab { side, reply } => {
            let _ = reply.send(backend.add_tab(side).await);
        }
        ApiCmd::CloseTab { side, id, reply } => {
            let _ = reply.send(backend.close_tab(side, id).await);
        }
        ApiCmd::SwitchTab { side, id, reply } => {
            let _ = reply.send(backend.switch_tab(side, id).await);
        }

        ApiCmd::GetPanelState { side, reply } => {
            let _ = reply.send(Ok(backend.read_state(side)));
        }

        ApiCmd::ReadFile { side, path, reply } => {
            let _ = reply.send(backend.read_file(side, path).await);
        }
        ApiCmd::StreamFile { side, path, reply } => {
            let _ = reply.send(backend.stream_file(side, path).await);
        }
        ApiCmd::WriteFile { side, path, content, reply } => {
            let _ = reply.send(backend.write_file(side, path, content).await);
        }
        ApiCmd::UploadFile { side, path, data, reply } => {
            let _ = reply.send(backend.upload_file(side, path, data).await);
        }

        ApiCmd::Delete { side, paths, reply } => {
            let _ = reply.send(Ok(()));
            let _ = backend.delete(side, paths).await;
        }
        ApiCmd::Mkdir { side, name, reply } => {
            let _ = reply.send(Ok(()));
            let _ = backend.mkdir(side, name).await;
        }
        ApiCmd::Rename { side, old_path, new_name, reply } => {
            let _ = reply.send(Ok(()));
            let _ = backend.rename(side, old_path, new_name).await;
        }
        ApiCmd::Copy { src_side, dst_side, paths, reply } => {
            let _ = reply.send(Ok(()));
            let _ = backend.copy(src_side, dst_side, paths).await;
        }
        ApiCmd::Move { src_side, dst_side, paths, reply } => {
            let _ = reply.send(Ok(()));
            let _ = backend.move_entries(src_side, dst_side, paths).await;
        }

        ApiCmd::GetDrives { reply } => {
            let _ = reply.send(Ok(backend.get_drives()));
        }
        ApiCmd::ActivateSource { side, key, reply } => {
            let _ = reply.send(backend.activate_source(side, key).await);
        }
        ApiCmd::GetConnections { reply } => {
            let _ = reply.send(Ok(backend.get_connections()));
        }
        ApiCmd::SaveConnection { connection, reply } => {
            let _ = reply.send(backend.save_connection(connection));
        }
        ApiCmd::DeleteConnection { name, reply } => {
            let _ = reply.send(backend.delete_connection(name));
        }
        ApiCmd::ConnectTo { side, connection, reply } => {
            let _ = reply.send(backend.connect_to(side, connection).await);
        }
        ApiCmd::RefreshPanel { side, reply } => {
            let _ = reply.send(backend.refresh_panel(side).await);
        }
        ApiCmd::GetSettings { reply } => {
            let _ = reply.send(backend.get_settings());
        }
        ApiCmd::SetSettings { values, reply } => {
            let _ = reply.send(backend.set_settings(values));
        }
        ApiCmd::GetViewerContent { reply } => {
            let _ = reply.send(backend.viewer_content());
        }
        ApiCmd::SetConnectionsDialog { open, reply } => {
            let _ = reply.send(backend.set_connections_dialog(open));
        }
        ApiCmd::ExportConnections { password, reply } => {
            let _ = reply.send(backend.export_connections(password));
        }
        ApiCmd::ImportConnections { data, password, reply } => {
            let _ = reply.send(backend.import_connections(data, password));
        }

        ApiCmd::ToggleFavorite { path, reply } => {
            backend.toggle_favorite(path);
            let _ = reply.send(Ok(()));
        }
        ApiCmd::GetFavoritesOnly { reply } => {
            let _ = reply.send(Ok(backend.get_favorites_only()));
        }
        ApiCmd::SetFavoritesOnly { value, reply } => {
            backend.set_favorites_only(value);
            let _ = reply.send(Ok(()));
        }

        other => return Some(other),
    }
    None
}
